arch/arm/src/am67/am67_irq.c

```diff
@@ -0,0 +1,379 @@
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/arch.h>
+#include <nuttx/irq.h>
+#include <assert.h>
+
+#include "arm_internal.h"
+#include "irq/irq.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define INTR_MAX_INTERRUPTS (512u)
+#define INTR_MAX_PRIORITY   (16u)
+#define INTRC_BASE_ADDR     0x2fff0000u
+
+#define VIM_BIT_POS(j)      ( (j) & 0x1fu )
+#define VIM_IRQVEC          (0x18u)
+#define VIM_ACTIRQ          (0x20u)
+#define VIM_RAW(j)          (0x400u + ((((j) >> 5) & 0xfu) * 0x20u))
+#define VIM_STS(j)          (0x404u + ((((j) >> 5) & 0xfu) * 0x20u))
+#define VIM_INT_EN(j)       (0x408u + ((((j) >> 5) & 0xfu) * 0x20u))
+#define VIM_INT_DIS(j)      (0x40cu + ((((j) >> 5) & 0xfu) * 0x20u))
+#define VIM_INT_MAP(j)      (0x418u + ((((j) >> 5) & 0xfu) * 0x20u))
+#define VIM_INT_TYPE(j)     (0x41cu + ((((j) >> 5) & 0xfu) * 0x20u))
+#define VIM_INT_PRI(j)      (0x1000u + ((j) * 0x4u))
+#define VIM_INT_VEC(j)      (0x2000u + ((j) * 0x4u))
+
+#define INTR_SUCCESS        ((int32_t)0)
+#define INTR_FAILURE        ((int32_t)-1)
+#define INTR_TIMEOUT        ((int32_t)-2)
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int32_t intr_get_irq(uint32_t *int_num);
+static void intr_ack_irq(uint32_t int_num);
+static void intr_clear_irq(uint32_t int_num);
+static void intr_set_irq_pri(uint32_t int_num, uint32_t priority);
+static uint32_t intr_get_irq_vec_addr(void);
+static void intr_set_irq_vec_addr(uint32_t int_num, uintptr_t vec_addr);
+
+static void utils_data_and_instruction_barrier(void);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: intr_get_irq
+ *
+ * Description:
+ *   Return the interrupt status corresponding to the given int_num.
+ *
+ ****************************************************************************/
+
+static int32_t intr_get_irq(uint32_t *int_num)
+{
+  volatile uint32_t *addr;
```

Comment:
Instead of modifying volatile pointers yourself, I would suggest using the `getreg32`/`putreg32` interfaces provided by NuttX. Keeps code consistent. Ditto for everywhere else you used this approach.